gitignore_tidy/core.py

```diff
@@ -1,123 +1,184 @@
+import dataclasses
+import itertools
 import pathlib
 import re
 import typing
+from functools import cached_property
 
-import typer
+from gitignore_tidy.logging import logger
 
 
 def tidy_file(path: pathlib.Path, *, allow_leading_whitespace: bool = False):
-
-    if not path.exists():
-        raise FileNotFoundError(f"{path} not found.")
-
-    with path.open("r") as f:
-        lines = f.read().splitlines()
-
+    lines = PlainLines.from_file(path)
     if len(lines) < 1:
-        typer.echo(f"{path} is empty.")
+        logger.info(f"File {path} is empty, not writing.")
         return
 
-    sorted_lines = _tidy_lines(
-        lines,
-        allow_leading_whitespace=allow_leading_whitespace,
-    )
-
-    _write_if_changed(sorted_lines, lines, path)
-
-
-def _normalize(lines: list[str], allow_leading_whitespace: bool = False) -> list[str]:
-
-    if not allow_leading_whitespace:
-        lines = [re.sub("^(!)? *\t*", "\\1", line) for line in lines]
-    lines = [re.sub(" *\t*$", "", line) for line in lines]
-    unique = []
-
-    for line in lines:
-        if line not in unique or line == "":
-            unique.append(line)
-    return unique
-
-
-def _sort_lines(lines: list[str]) -> list[str]:
-    original = lines
-    non_negated = [re.sub("^!", "", line) for line in lines]
-    sorted_non_negated = sorted(non_negated)
-    sorted_negated = []
-    for line in sorted_non_negated:
-        negated_cand = "!" + line
-        if negated_cand in original:
-            sorted_negated.append(negated_cand)
-        else:
-            sorted_negated.append(line)
-
-    return sorted_negated
-
+    tidy_plain_lines = tidy_lines(lines, allow_leading_whitespace=allow_leading_whitespace)
+    if lines.lines == tidy_plain_lines.lines:
+        logger.info(f"{path} already tidy.")  # TODO use logger module
+    else:
+        tidy_plain_lines.to_file(path)
+        logger.info(f"Successfully written {path}.")
+
+
+def tidy_lines(lines: "PlainLines", allow_leading_whitespace: bool) -> "PlainLines":
+    normalised_contents = lines.normalize(allow_leading_whitespace=allow_leading_whitespace)
+    sorted_sections = Sections(tuple(section.sort() for section in normalised_contents.split()))
+    return sorted_sections.as_plain()
+
+
+@dataclasses.dataclass(frozen=True)
+class PlainLines:
+    """
+    Flat representation of a `.gitignore` file or parts of it
+    """
+
+    lines: list[str]
```

Comment:
I was also wondering, but more about using a concrete datatype for annotation, i.e. tuple here. Sometimes I am not sure if using generic containers like sequence is more suited than specific ones. I read trade-off involve how abstract you want your codebase to be (easier to refactor / adapt with high abstraction) and what other conventions you find as well as how people use your library (i.e. it might be convenient to provide abstract annotation so people can supply different data structures that inherit from the abc). More ideas? I'll try to make the switch to tuple now and maybe later to `Sequence`.

Reply:
I try to be as generic a possible with argument and/or dataclass field types, but specific with return types. That gives library users the most flexibility.